File: apps/mobile-app/android/app/src/main/java/net/aliasvault/app/vaultstore/repositories/BaseRepository.kt
package net.aliasvault.app.vaultstore.repositories

import net.aliasvault.app.utils.DateHelpers
import net.aliasvault.app.vaultstore.AppError
import net.aliasvault.app.vaultstore.VaultDatabase
import java.util.UUID

/**
 * Base repository class with common database operations.
 */
open class BaseRepository(
    /** The database component used for executing queries. */
    protected val database: VaultDatabase,
) {
    /**
     * The manifest new rows outside any folder or item are stamped with: the personal manifest. Rows inside a
     * folder or item take that parent's manifest through the SQL instead. Throws when no manifest has been
     * recorded yet, since an unstamped row would make every later push of the vault fail.
     */
    protected fun activeManifestId(): String {
        return database.getPersonalManifestId() ?: throw AppError.ManifestNotRecorded()
    }

    /**
     * The manifest a manifest-scoped row belongs to, looked up from the row itself: the personal manifest when
     * the row exists there, else the lowest manifest id holding it. Null when no such row exists.
     */
    protected fun resolveRowManifestId(table: String, id: String, column: String = "Id"): String? {
        val rows = executeQuery("SELECT ManifestId FROM $table WHERE $column = ? ORDER BY ManifestId", arrayOf(id))
        val manifestIds = rows.mapNotNull { it["ManifestId"] as? String }
        if (manifestIds.isEmpty()) return null
        val personal = database.getPersonalManifestId()
        return manifestIds.firstOrNull { it == personal } ?: manifestIds.first()
    }

    /**
     * The grouping key of a manifest-scoped row, for joining rows of one query to rows of another in memory.
     */
    protected fun scopedKey(manifestId: String, id: String): String {
        return "${manifestId.lowercase()}${id.lowercase()}"
    }

    // MARK: - Transaction Helpers

    /**
     * Execute a function within a transaction. The commit persists the vault and marks it dirty, so every
     * repository write reaches the next sync.
     * @param operation The function to execute within the transaction
     * @return The result of the function
     */
    fun <T> withTransaction(operation: () -> T): T {
        database.beginTransaction()
        return try {
            val result = operation()
            database.commitTransaction()
            result
        } catch (e: Exception) {
            database.rollbackTransaction()
            throw e
        }
    }

    // MARK: - Utility Methods

    /**
     * Generate a new id.
     * @return A new UUID string
     */
    fun generateId(): String {
        return UUID.randomUUID().toString()
    }

    /**
     * Get the current timestamp in the standard format.
     * @return Current timestamp string
     */
    fun now(): String {
        return DateHelpers.now()
    }

    // MARK: - Database Operation Helpers

    /**
     * Execute a SELECT query on the database.
     */
    protected fun executeQuery(query: String, params: Array<Any?>): List<Map<String, Any?>> {
        return database.query(query, params.toList())
    }

    /**
     * Execute an UPDATE, INSERT, or DELETE query on the database.
     */
    protected fun executeUpdate(query: String, params: Array<Any?>): Int {
        return database.execute(query, params.toList())
    }
}

File: apps/mobile-app/ios/VaultStoreKit/Database/BaseRepository.swift
import Foundation
import VaultUtils

/// The grouping key of a manifest-scoped row (same id can exist in several manifests by design).
internal func scopedKey(manifestId: String, id: String) -> String {
    return manifestId + id
}

/// Base repository class with common database operations.
public class BaseRepository {
    /// The database client used for executing queries.
    internal let client: DatabaseClient

    /// Initialize the repository with a database client.
    /// - Parameter client: The database client to use
    public init(client: DatabaseClient) {
        self.client = client
    }

    /// The manifest new rows outside any folder or item are stamped with: the personal manifest. Rows inside a
    /// folder or item take that parent's manifest through the SQL instead. An unrecorded manifest is an error,
    /// never the empty (unstamped) scope, which every later push would reject.
    public func writeManifestId() throws -> String {
        guard let manifestId = client.personalManifestId(), !manifestId.isEmpty else {
            throw AppError.manifestNotRecorded
        }
        return manifestId
    }

    // MARK: - Transaction Helpers

    /// Execute a function within a transaction.
    /// Automatically handles begin, commit, and rollback.
    /// - Parameter operation: The function to execute within the transaction
    /// - Returns: The result of the function
    public func withTransaction<T>(_ operation: () throws -> T) throws -> T {
        try client.beginTransaction()
        do {
            let result = try operation()
            try client.commitTransaction()
            return result
        } catch {
            try? client.rollbackTransaction()
            throw error
        }
    }

    // MARK: - Utility Methods

    /// Generate a new id.
    /// - Returns: A new UUID string
    public func generateId() -> String {
        return UUID().uuidString.lowercased()
    }

    /// Get the current timestamp in the standard format.
    /// - Returns: Current timestamp string
    public func now() -> String {
        return DateHelpers.now()
    }
}
